### DISCUSSION TITLE
[General] <Greeter web dApp, can't setGreeting> zkgreeter.j.pl

### DISCUSSION THREAD

**Discussion by 0xjaqbek** [zkSync-Community-Hub/zksync-developers#179] [General] [answered]
### Environment

Mainnet

### Provide a brief description of the functionality you're trying to implement.

www.zkgreeter.j.pl
Builded website connects to wallet and enables to set Greeting on deployed contract. It's also able to read written Greetings.

### What is the specific issue or error you're encountering?

After succesfully connecting wallet i'm able to read written greetings by users or owners Greeting but i can't set new Greeting. 

### Can you share the error messages or logs you're receiving, if any?

From browsers console after connecting wallet:
Error setting user's greeting: Error: zkSync API response error: code -32602; message: Invalid params: invalid value: map, expected map with a single key.
    r https://unpkg.com/zksync@0.12.0/dist/main.js:1
    request https://unpkg.com/zksync@0.12.0/dist/main.js:1
    t https://unpkg.com/zksync@0.12.0/dist/main.js:1
[zkgreeter.j.pl:191:21](http://zkgreeter.j.pl/)
    setUserGreeting http://zkgreeter.j.pl/:191

### Have you made any recent changes to the contract before encountering this issue?

No. Contract is working fine when interacted thru blockexplorer.

### Are there any external libraries or contracts that your contract interacts with?

It's Greeter so no.

### Can you provide the relevant portions of your contract code where the issue is occurring?

It's related to (script on) website that interacts with contract and its that function:

```
// Function to handle wallet connection and set user's greeting
async function setUserGreeting() {
    if (window.ethereum) {
        try {
            // Request access to the user's Ethereum wallet
            const accounts = await window.ethereum.request({ method: 'eth_requestAccounts' });
            console.log("Ethereum accounts obtained:", accounts);

            // Prompt the user for a new greeting
            const newGreeting = prompt("Enter a new user's greeting:");
            if (newGreeting !== null) {
                // Set the user's greeting using zkSync.js
                const syncProvider = await zksync.Provider.newHttpProvider("https://api.zksync.io/jsrpc");
                const ethersProvider = new ethers.providers.Web3Provider(window.ethereum);
                const signer = ethersProvider.getSigner();
                const syncWallet = await zksync.Wallet.fromEthSigner(signer, syncProvider);

                const changePubkey = await syncWallet.setSigningKey({feeToken: 0}); // Use 0 for ETH
                await changePubkey.awaitReceipt();

                const nonce = await syncWallet.getNonce("committed");
                const fee = await syncProvider.getTransactionsBatchFee(['Transfer'], [syncWallet.address], 'ETH');

                const tx = {
                    to: "0xf0CF19F798BF5B98b5201906EC3a460075612C83",
                    token: "ETH",
                    amount: ethers.utils.parseEther("0"), // 0 ETH
                    fee,
                    nonce,
                    validFrom: 0,
                    validUntil: Math.floor(Date.now() / 1000) + 3600, // 1 hour from now
                    data: ethers.utils.defaultAbiCoder.encode(['string'], [newGreeting]) // encoding the function call
                };

                const signedTx = await syncWallet.createSignedContractCall(tx);
                const receipt = await syncProvider.submitTx(signedTx);

                console.log("Transaction receipt:", receipt);

                console.log("User's greeting set successfully.");
                alert("User's greeting set successfully.");
            }
        } catch (error) {
            console.error("Error setting user's greeting:", error);
            alert("Error setting user's greeting.");
        }
    } else {
        alert("No Ethereum wallet detected. Please install MetaMask or another Ethereum wallet.");
    }
}
```

### Have you tried to isolate the problem, and if so, what were the results?

nope.

### What steps have you already taken to try to resolve the issue?

Changed that script on website.

**Comment by vasyl-ivanchuk** [comment, marked as answer]
Hey @mironikson 👋 ,

We have a comprehensive [Hello world tutorial](https://era.zksync.io/docs/dev/building-on-zksync/hello-world.html) that provides in-depth guidance on the creation, deployment, and verification of a Greeter contract. Additionally, it outlines the steps to interact with the contract from the frontend dApp. I encourage you to check it out as it aligns with what you are trying to achieve.

Bellow you'll find some code snippets from the tutorial, showing how to interact with the deployed contract:

```
initializeProviderAndSigner() {
    this.provider = new Provider('https://testnet.era.zksync.dev');
    this.signer = (new Web3Provider(window.ethereum)).getSigner();
    this.contract = new Contract(
        GREETER_CONTRACT_ADDRESS,
        GREETER_CONTRACT_ABI,
        this.signer
    );
},
async getGreeting() {
    // Smart contract calls work the same way as in `ethers`
    return await this.contract.greet();
},
async changeGreeting() {
    // ...
    const txHandle = await this.contract.setGreeting(this.newGreeting, await this.getOverrides());
    await txHandle.wait();
    // ...
},
```

**Comment by andybrierton** [comment]
It seems like the issue might be related to the format of the data being sent to the contract. The error message "Invalid params: invalid value: map, expected map with a single key" suggests that the data format might not be correct.

Make sure that the data being encoded using ethers.utils.defaultAbiCoder.encode is in the expected format for the contract function you are calling. Double-check the ABI of the contract function to ensure that the encoding matches the expected parameters.

Additionally, ensure that the contract address and token are correct, and that the user has sufficient funds to cover the transaction fee.

You may also consider debugging by logging intermediate values and checking for any discrepancies between the expected and actual values during the transaction process.

If the issue persists, consider reaching out to the ZKSync community or support for further assistance.

**Comment by clevelandmarcus** [comment]
The error message indicates that there's an issue with the parameters being passed to the zkSync API when setting the user's greeting. Check the format of the parameters being sent and ensure they match the expected format by the API.

**Comment by sundaramjacob** [comment]
The error "Invalid params: invalid value: map, expected map with a single key" suggests that there might be an issue with the format of parameters being sent to zkSync API. Check the data being passed to the API and ensure it conforms to the expected format, particularly when setting the user's greeting.
